Support Py2 and Py3 by switching to Pybind11

nilsvu opened the pull request:
## Proposed changes

To support both Python2 and Python3 I found it easier to ditch boost.python and switch to [pybind11](https://pybind11.readthedocs.io/en/stable/) (see https://github.com/sxs-collaboration/spectre/issues/1829 for a brief comparison). By doing so this PR accomplishes the following:

- Python 2 and Python 3 support for our bindings by selecting the `PYTHON_EXECUTABLE` in CMake. Since pybind11 is header-only, there is no need to compile different versions of boost.
- Closes #1518.
- The bindings compile on macOS. Closes #1923.
- Our CI tests the bindings with Py2 and Py3.
- We can make our data structures compatible with Numpy arrays now. This PR includes a few examples (I can factor them into separate PRs if the reviewers prefer):
  - `DataVector` and `Matrix` can be converted to and from Numpy arrays (with and without copying).
  - Functions that take and return `double`s can be made to work with Numpy arrays using `py::vectorize`.
- [ ] Closes #1839. (@wthrowe please confirm)
- Closes #1829.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [x] Build system
- [x] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

moxcodes commented:
I think it's worth discussing our stance regarding this feature and the impending python 2 end of life. Thoughts?

nilsvu commented:
Yes we should discuss that. Before this PR, everything was tested on Py2. We could either switch to Py3 exclusively, or test on both Py2 and Py3. I think supporting both can help on older clusters.

nilsdeppe commented:
Unfortunately I don't think it'll be possible to require python 3 for another several years. The reason being is we realistically need to support supercomputers that are up to about 5 years old. Getting a working python installation is rather difficult and so I think we need to be flexible still for the next few years to be on the safe side. It'll be easier to maintain support for both 2 & 3 than have to add support for 2 in a few years if we get thrown onto a machine that doesn't have 3. That's my concern and opinion :man_shrugging:

moxcodes commented:
Okay, I think that's pretty convincing. I do think that there will tend to be a point where it's worth dropping support for older compilers and interpreters, because until we do, we don't get to access to the full features of newer versions and 'end of life' seems on the surface a good time to discuss it. But my understanding of scientific computing realities is in agreement with what you're saying, that it is unlikely that every package and cluster is going to make the upgrade to python 3 in the next two months...

kidder commented on src/DataStructures/Python/Numpy.hpp:
```diff
@@ -6,9 +6,20 @@
+#ifdef SPECTRE_NUMPY_IMPORT_ARRAY
+#include <numpy/arrayobject.h>
+#include "PythonBindings/NumpyImportArrayWrapper.hpp"
+#else
```
add `// SPECTRE_NUMPY_IMPORT_ARRAY` to `else` and `endif` statements as it helps readability

kidder commented on src/PythonBindings/NumpyImportArrayWrapper.hpp:
```diff
@@ -0,0 +1,20 @@
+spectre_numpy_import_array() {
+  import_array();
+#if PY_MAJOR_VERSION >= 3
+  return nullptr;
```
can you always return `nullptr` ?

kidder commented on tests/Unit/Pypp/SetupLocalPythonEnvironment.cpp:
```diff
@@ -15,6 +15,9 @@
 #include <pydebug.h>
 
+// Must include this after `numpy/arrayobject.h`
+#include "PythonBindings/NumpyImportArrayWrapper.hpp"
```
inline comment and group below with other includes

kidder commented on CMakeLists.txt:
```diff
@@ -12,6 +12,16 @@ site_name(HOSTNAME)
 
 list(APPEND CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/cmake")
 
+# First search for Python3 then fall back to Python2, since specifying
```
put this in `cmake/SetupPython.cmake`  and also change `SpectreSetupPython` to `SetupSpectrePythonPackage`

kidder commented on containers/Dockerfile.buildenv:
```diff
@@ -23,6 +23,12 @@ RUN apt-get update -y \
     && apt-get install -y software-properties-common \
     && add-apt-repository ppa:ubuntu-toolchain-r/test
 
+# Add Kitware repo for cmake
```
let's just divide python2 and python3 tests into separate Travis build targets and only test a single one as we need to make sure the boost python module is consistent with the spectre ones

nilsdeppe commented:
@moxcodes yes, I agree. This issue somewhat tracks systems we're using and what they offer for compiler support: https://github.com/sxs-collaboration/spectre/issues/442 We could have an issue (or add to that table) python version support and then once everything is safely moved over to python3 we can remove python2 support. What do you think?

moxcodes commented:
@nilsdeppe I think that's a good suggestion, to have another column indicating python version support.

nilsvu committed:
Install Py3 packages in container

nilsvu committed:
Rename SpectreSetupPython to SpectreSetupPythonPackage

nilsvu committed:
Add label for Python tests

nilsvu committed:
Add Py3 instructions to documentation

nilsvu committed:
Install Pybind11 in Docker container

nilsvu committed:
Add Pybind11 dependency for building Python bindings

nilsvu committed:
Convert PyDataStructures to Pybind11

nilsvu committed:
Add Numpy-interoperability to DataVector

nilsvu committed:
Add Numpy interoperability to Matrix

nilsvu committed:
Convert PyInformer to Pybind11

nilsvu committed:
Convert PyH5 to Pybind11

nilsvu committed:
Convert PyEquationsOfState to Pybind11

nilsvu committed:
Convert PyGeneralRelativitySolutions to Pybind11

nilsvu committed:
Vectorize TOV Python wrappers for Numpy

nilsvu committed:
Add CI tests for Py2 and Py3

nilsvu commented:
@kidder @nilsdeppe @moxcodes This PR is now ready to review! I pushed a lot of small commits, but delete more code than I'm adding :)

@nilsdeppe could you take a look at the Dockerfile and update the container if you're happy with the changes? I suggest we keep boost.python in the container for a while before removing it :)

nilsdeppe commented on CMakeLists.txt:
```diff
@@ -72,7 +72,7 @@ include(SetupLibsharp)
 
 include(SetupLIBCXXCharm)
-include(SpectreSetupPython)
+include(SetupSpectrePythonPackage)
```
Please don't do this rename

nilsvu replied:
agreed on `SpectreSetupPythonPackage` in a call

nilsdeppe commented on containers/Dockerfile.buildenv:
```diff
@@ -65,11 +65,14 @@ RUN apt-get update -y \
-    && pip install numpy scipy matplotlib pandas h5py jupyterlab \
-    && pip install coverxygen beautifulsoup4 pybtex
+    && pip2 install numpy scipy h5py yapf \
+    && pip3 install numpy scipy h5py yapf \
```
I don't think we need all of these, only h5py

nilsleiffischer replied:
note from call: add comment for reasoning

nilsvu replied:
note from call: add comment for reasoning

nilsdeppe commented on containers/Dockerfile.buildenv:
```diff
@@ -101,6 +101,11 @@ RUN wget https://bitbucket.org/blaze-lib/blaze/downloads/blaze-3.2.tar.gz -O bla
     && mv catch.hpp /usr/local/include \
+    && git clone https://github.com/pybind/pybind11.git \
+    && cd pybind11 && mkdir build && cd build \
+    && cmake -D PYBIND11_TEST=OFF .. && make install \
```
I don't like this install. If we can't just move the include directory to the desired location we can't really use pybind11... Furthermore, this should use a release.

nilsvu replied:
note from call: copy pybind11 cmake tools into `external` so we only need to `mv` the `include` directory

nilsdeppe commented on src/DataStructures/Python/Bindings.cpp:
```diff
@@ -1,35 +1,16 @@
-// 2. When C++ modules land this type of pattern will no longer really be
-//    possible since it goes against exactly what modules is trying to do: make
-//    things modular.
+namespace py = pybind11;
```
please use the `bp` alias instead

kidder replied:
what is `bp` ?

nilsvu replied:
we agreed to make this `py` in the call. `bp` was `boost::python`

nilsdeppe reviewed (changes requested):
I can't review this until you make the commits easier to review. For example, the convert DataStructures commit makes a whole host of changes that make it annoying to review.

nilsdeppe commented on src/DataStructures/Python/DataVector.cpp:
```diff
@@ -27,6 +27,41 @@ void bind_datavector(py::module& m) {
+             if (info.ndim != 1) {
+               throw std::runtime_error("Incompatible dimension.");
+             }
+            const size_t size = static_cast<size_t>(info.shape[0]);
```
Format this?

nilsdeppe commented on src/DataStructures/Python/Matrix.cpp:
```diff
@@ -20,8 +20,38 @@ namespace py_bindings {
 
```
Please update commit message: DataVector -> Matrix

nilsdeppe reviewed (commented):
Review over a call with @nilsleiffischer.

I'm happy for you to squash everything

kidder reviewed (changes requested):
I have no requested changes other than that this is tested once it passes CI on:
- [ ] wheeler
- [ ] frontera
- [x] @kidder macbook 
- [ ] @nilsdeppe laptop
- [x] @wthrowe home desktop

nilsvu committed:
Update Pybindings docs

nilsvu commented:
Done @nilsdeppe. Please take a look at the Dockerfile and update the container if you're happy with the changes. Then you can give the new bindings a try on your machines :)

Edit: These are the two relevant commits that update the Dockerfile: 0ee1726a0cc34d8cd9791054764440192d5b4001 and eb2d23e3cc559e5e958c298a85bb1efff9bdbaa5

nilsdeppe commented on containers/Dockerfile.buildenv:
```diff
@@ -101,10 +101,8 @@ RUN wget https://bitbucket.org/blaze-lib/blaze/downloads/blaze-3.2.tar.gz -O bla
     && rm -rf brigand \
     && wget https://github.com/catchorg/Catch2/releases/download/v2.2.1/catch.hpp -O catch.hpp \
     && mv catch.hpp /usr/local/include \
-    && git clone https://github.com/pybind/pybind11.git \
```
I think this got squashed into the wrong commit, or you can remove the previous commit

nilsvu replied:
yes I squashed it into the wrong commit. sorry for not catching that. should be fixed now.

nilsdeppe commented on tools/FileTestDefs.sh:
```diff
@@ -298,6 +298,10 @@ license() {
               'docs/config/header.html' \
               'docs/config/layout.xml' \
               'docs/config/MathJax.js$' \
+              'external/pybind11/tools/FindPythonLibsNew.cmake' \
```
Can you try if the regex `'external/*` works?

nilsvu replied:
looks like it does 👍

nilsdeppe commented on containers/Dockerfile.buildenv:
```diff
@@ -65,11 +65,14 @@ RUN apt-get update -y \
 # Install Python packages
+# We only install packages that are needed by the build system (e.g. to compile
+# Python bindings or build documentation) or used by Python code that is
+# unit-tested. Any other packages can be installed on-demand.
```
Could you specify that `We need coverxygen, beatifulsoup4, and pybtex for both python2 and python3 so that people can build the documentation with either python 2 or 3.`

nilsdeppe reviewed (commented):
Looks good, couple minor things. Please squash :)

nilsvu committed:
Bundle pybind11 cmake files

We can't rely on these files being installed with the headers on
all computers, so we bundle them.

nilsvu committed:
Update setup.py

nilsdeppe commented:
Looks good! I'm updating the container now and I'll restart the tests once the container is pushed

nilsdeppe commented on containers/Dockerfile.buildenv:
```diff
@@ -108,6 +108,9 @@ RUN wget https://bitbucket.org/blaze-lib/blaze/downloads/blaze-3.2.tar.gz -O bla
     && wget https://github.com/catchorg/Catch2/releases/download/v2.2.1/catch.hpp -O catch.hpp \
     && mv catch.hpp /usr/local/include \
+    && wget https://github.com/pybind/pybind11/archive/v2.4.3.tar.gz -O pybind11.tar.gz \
+    && mv pybind11/include/pybind11 /usr/local/include \
```
These should be:

```
    && wget https://github.com/pybind/pybind11/archive/v2.4.3.tar.gz -O pybind11.tar.gz \
    && tar xzf pybind11.tar.gz \
    && mv pybind11-2.4.3/include/pybind11 /usr/local/include \
    && rm -rf pybind11 \
```

nilsvu replied:
🤦‍♂️ I should have built the container to check. Doing so now

nilsdeppe replied:
I've pushed the updated container built with exactly the lines I posted.

nilsdeppe commented:
@nilsleiffischer clang-tidy is failing

nilsvu committed:
Remove boost.python dependency

nilsvu committed:
Inline SPECTRE_LINK_PYBINDS, making it work on macOS

nilsvu commented:
Fixed a bunch of clang-tidy issues and added `NOLINT`s for pybind11-specific clang-tidy complaints. This is passing the CI checks now, running in the updated container. @kidder @nilsdeppe @wthrowe could you do the respective tests that @kidder suggested?

nilsdeppe commented:
A clean build doesn't work on my machine:
```
Test project /home/nils/SpECTRE/spectre/build_clangpy3
    Start 1056: Unit.Visualization.Python.GenerateXdmf
1/1 Test #1056: Unit.Visualization.Python.GenerateXdmf ...***Failed    0.20 sec
Traceback (most recent call last):
  File "/home/nils/SpECTRE/spectre/tests/Unit/Visualization/Python/Test_GenerateXdmf.py", line 6, in <module>
    import spectre.Informer as spectre_informer
  File "/home/nils/SpECTRE/spectre/build_clangpy3/bin/python/spectre/Informer/__init__.py", line 1, in <module>
    from ._PyInformer import *
ImportError: /home/nils/SpECTRE/spectre/build_clangpy3/lib/libInformer.so: undefined symbol: _Cmi_numpes
```
and
```
Test project /home/nils/SpECTRE/spectre/build_clangpy3
    Start 1054: Unit.AnalyticSolutions.GeneralRelativity.Python.Tov
1/1 Test #1054: Unit.AnalyticSolutions.GeneralRelativity.Python.Tov ...***Failed    0.04 sec
Traceback (most recent call last):
  File "/home/nils/SpECTRE/spectre/tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Python/Test_Tov.py", line 4, in <module>
    import spectre.PointwiseFunctions.AnalyticSolutions.GeneralRelativity \
  File "/home/nils/SpECTRE/spectre/build_clangpy3/bin/python/spectre/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/__init__.py", line 1, in <module>
    from ._PyGeneralRelativitySolutions import *
ImportError: /home/nils/SpECTRE/spectre/build_clangpy3/lib/libSpectral.so: undefined symbol: sharp_make_gauss_geom_info
```

wthrowe commented:
Works on my machine.  (One test fails, but it's unrelated: #1957.)

I'll test again once changes are made for @nilsdeppe's problem.

wthrowe commented:
Also, this does not fix #1839.

kidder commented:
@wthrowe can you clarify about #1839.  I find `Python` at the top of my `Label Time Summary` when I run `ctest` (not sure why it is the only label that isn't all lower-case...)

nilsvu commented:
@nilsdeppe could you post the command you invoked cmake with, and which targets you built and ran? Did the other Python tests succeed, eg DataStructures and H5? If I can‘t reproduce your issue with that information, it would be helpful if you could send me more info about you environment (eg CMake cache) on slack

wthrowe commented:
You're right, #1839 is fixed.  I only checked where I expected it to be alphabetically.

nilsdeppe commented:
@nilsleiffischer 
```
cmake -GNinja -D CHARM_ROOT=/home/nils/SpECTRE/charm_68/multicore-linux64-clang -D CMAKE_C_COMPILER=clang -D CMAKE_CXX_COMPILER=clang++ -D ASAN=NO -D BUILD_SHARED_LIBS=ON -D DEBUG_SYMBOLS=OFF -D CMAKE_BUILD_TYPE=Debug -D BUILD_PYTHON_BINDINGS=ON ..
```

I built RunTests and test-executables. It was `Tov.Python`, `Informer.Python`, `VolumeData.python`, and `GenerateXdmf` that failed. 3 with the _Cmi_num_pes

nilsvu commented:
@nilsdeppe have you tried building with static libraries? And do you know if the pybindings on current `develop` work with shared libs? I’m having trouble reproducing your issue because shared libs don’t work in the container currently (probably due to yamlcpp not installed with shared libs, I’ll try to work around that). I’m also going to try on another system such as minerva.

nilsdeppe commented:
Building with static libs I no longer get the issue. I do think we need to support dynamic libs because they help reduce file size a lot.

nilsdeppe commented:
Okay, it looks like the correct solution is that SpECTRE libraries that call into external libraries, e.g. HDF5, libsharp, LAPACK, BLAS, etc. need to mark those libs as INTERFACE link targets.

nilsvu commented:
@nilsdeppe Agreed. The current boost.python bindings also don't link correctly with shared libs, opened #1961. Also, building with shared libs currently doesn't work in the container, so I fixed that in #1962. Then I can work on getting shared libs linked correctly with the bindings once the container supports it. Does that work for you?

@kidder could you verify on wheeler (and perhaps frontera) to complete your list in https://github.com/sxs-collaboration/spectre/pull/1782#pullrequestreview-349740503?

nilsdeppe commented:
@nilsleiffischer sure, sounds good to me.

nilsdeppe commented:
On wheeler with Clang I get:
```
8/9 Test #1056: Unit.AnalyticSolutions.GeneralRelativity.Python.Tov ...***Failed    0.30 sec
Traceback (most recent call last):
  File "/home/nils/spectre_nils/tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Python/Test_Tov.py", line 4, in <module>
    import spectre.PointwiseFunctions.AnalyticSolutions.GeneralRelativity \
  File "/home/nils/spectre_nils/build_clang/bin/python/spectre/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/__init__.py", line 1, in <module>
    from ._PyGeneralRelativitySolutions import *
ImportError: /home/nils/spectre_nils/build_clang/bin/python/spectre/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/_PyGeneralRelativitySolutions.so: undefined symbol: _ZTVN10__cxxabiv120__si_class_type_infoE
```

With GCC things work fine.

Can you add module load/unloads to the Wheeler envs for pybind11/2.4.3?

nilsvu committed:
Load Pybind11 on Wheeler

Co-authored-by: Nils Deppe <nd357@cornell.edu>

nilsvu commented:
@nilsdeppe done

nilsdeppe commented:
I think the correct solution is:

```
index 5b813ce..0658693 100644
--- a/cmake/SpectreSetupPythonPackage.cmake
+++ b/cmake/SpectreSetupPythonPackage.cmake
@@ -114,6 +114,11 @@ function(SPECTRE_PYTHON_ADD_MODULE MODULE_NAME)
       LIBRARY_OUTPUT_NAME "_${ARG_LIBRARY_NAME}"
       LIBRARY_OUTPUT_DIRECTORY ${MODULE_LOCATION}
       )
+    set_target_properties(
+      ${ARG_LIBRARY_NAME}
+      PROPERTIES
+      LINK_FLAGS ${CMAKE_CXX_LINK_FLAGS}
+      )
     set(SPECTRE_PYTHON_MODULE_IMPORT "from ._${ARG_LIBRARY_NAME} import *")
     add_dependencies(test-executables ${ARG_LIBRARY_NAME})
   endif(BUILD_PYTHON_BINDINGS AND NOT "${ARG_SOURCES}" STREQUAL "")
```

and then have `-lstdc++` specified in the `wheeler_clang.sh` as the `CMAKE_CXX_LINK_FLAGS`

nilsvu committed:
Pass link flags to Python modules

Fixes an issue on Wheeler

Co-authored-by: Nils Deppe <nd357@cornell.edu>

nilsvu commented:
Thanks @nilsdeppe. Could you try again on wheeler when you get the chance? I don't have an account yet.

kidder commented:
works on all my machines; we haven't even tried compiling develop on frontera yet so we'll not hold up this PR for testing there...

nilsdeppe reviewed (commented):
Okay, this works on Wheeler now. :+1:

@wthrowe could you double check this on your machine please? Thanks!

nilsvu commented:
Fixed an issue with commit metadata not showing up on GitHub (I had messed up my `user.email` gitconfig) and rebased to make sure the CI runs again on current develop.

@wthrowe please test on your machine when you get the chance. thanks for reviewing @nilsdeppe @kidder @wthrowe !

wthrowe commented:
Works for me.  (Except for #1957.)
